crates/services/consensus_module/poa/src/sync.rs

```diff
@@ -0,0 +1,563 @@
+        // Nothing to shut down because we don't have any temporary state that should be dumped,
+        // and we don't spawn any sub-tasks that we need to finish or await.
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum InnerSyncState {
+    /// We are not connected to at least `min_connected_reserved_peers` peers.
+    ///
+    /// InsufficientPeers -> SufficientPeers
+    InsufficientPeers(BlockHeight),
+    /// We are connected to at least `min_connected_reserved_peers` peers.
+    ///
+    /// SufficientPeers -> Synced(...)
+    SufficientPeers(BlockHeight),
+    /// We can go into this state if we didn't receive any notification
+    /// about new block height from the network for `time_until_synced` timeout.
+    ///
+    /// We can leave this state only in the case, if we received a valid block
+    /// from the network with higher block height.
+    ///
+    /// Synced -> either InsufficientPeers(...) or SufficientPeers(...)
+    Synced(BlockHeight),
+}
+
+impl InnerSyncState {
+    fn from_config(
+        min_connected_reserved_peers: usize,
+        time_until_synced: Duration,
+        block_height: BlockHeight,
+    ) -> Self {
+        match (min_connected_reserved_peers, time_until_synced) {
+            (0, Duration::ZERO) => InnerSyncState::Synced(block_height),
+            (0, _) => InnerSyncState::SufficientPeers(block_height),
+            _ => InnerSyncState::InsufficientPeers(block_height),
+        }
+    }
+
+    fn change_state_on_block(&mut self, new_block_height: BlockHeight) -> bool {
+        let mut state_changed = false;
+        let current_block_height = self.block_height();
+
+        if &new_block_height > current_block_height {
+            match self {
+                InnerSyncState::InsufficientPeers(_) => {
+                    *self = InnerSyncState::InsufficientPeers(new_block_height);
+                }
+                InnerSyncState::SufficientPeers(_) => {
+                    *self = InnerSyncState::SufficientPeers(new_block_height);
+                }
+                InnerSyncState::Synced(_) => {
+                    // we considered to be synced but we're obiously not!
+                    *self = InnerSyncState::SufficientPeers(new_block_height);
+                    state_changed = true;
+                }
+            }
+        }
+
+        state_changed
+    }
+
+    fn change_state_on_peers_update(
+        &mut self,
+        currently_connected_peers: usize,
+        min_connected_reserved_peers: usize,
+    ) -> bool {
+        let sufficient_peers = currently_connected_peers >= min_connected_reserved_peers;
+
+        match self {
+            InnerSyncState::InsufficientPeers(block_height) if sufficient_peers => {
+                *self = InnerSyncState::SufficientPeers(*block_height);
+                true
+            }
+            InnerSyncState::SufficientPeers(block_height)
+            | InnerSyncState::Synced(block_height)
```

Comment:
Hmm, I think if we started producing blocks, then we should continue to do that even with the insufficient number of peers.